### PR TITLE
core: handle lookup paths being symlinks

### PR DESCRIPTION
**Pull request by andir**
With a recent change paths leaving the statically known lookup paths
would be treated differently then those that remained within those. That
was done (AFAIK) to consistently handle alias names. Unfortunately that
means that on some distributions, especially those where /etc/ consists
mostly of symlinks, would trigger that new detection for every single
unit in /etc/systemd/system. The reason for that is that the units
directory itself is already a symlink.

This is related to the issue discussed in #20003. cc @keszybz.

I haven't fully analyzed all the changes that went into the original change but this change seems to do what is required to make this edge case work (again). Let me know if there is another approach I could take. Unfortunately havin /etc/systemd/system *not* as a symlink isn't workable.

### CODE REVIEW

File: src/basic/unit-file.c
Comment (andir) [inline src/basic/unit-file.c]: I could probably drop this debug message here. Not sure why I left this in. Not really adding much information in this error case IMO.

**Comment by andir** [comment]
@yuwata thanks for adding the labels. Is there anyone I could ping for a review?

File: src/basic/unit-file.c
Comment (poettering) [inline src/basic/unit-file.c]: hmm, we usually break our text at 109 chars, see .vimrc and .dir-locals.el

File: src/basic/unit-file.c
Comment (poettering) [inline src/basic/unit-file.c]: why include both the regular and the "expanded" ones? the latter should suffice, no?

File: src/basic/unit-file.c
Comment (poettering) [inline src/basic/unit-file.c]: hmm, so why not just iterate through the reduced search path here?

i.e it appears to me the "expanded" search path should be called the "reduced" search path, and also maybe we should call strv_uniq() or so, to maybe strip duplicates. And then after all that we shoud operate purely on that and basically forget about original search path while building the map.

**Review by poettering — CHANGES_REQUESTED** [review]
i think @keszybz should also have a closer look, given he touched this code last

File: src/basic/unit-file.c
Comment (andir) [reply src/basic/unit-file.c]: You are right. That part is missing from the editorconfig. I prefer using that over sourcing "random" external vimrc's. I've filed a PR (https://github.com/systemd/systemd/pull/20894) to add the changes to the editorconfig and will update this PR later.

File: src/basic/unit-file.c
Comment (andir) [reply src/basic/unit-file.c]: >hmm, so why not just iterate through the reduced search path here?

I don't really know anymore. IIRC I was just proposing the minimal solution here to get some feedback on the approach. You might be correct that we can change all the `search_path` usages to the new list of paths. My only concern would be that in cases where e.g. `/etc/systemd/system` is already a symlink to another location we would then never see that path in `systemctl {status,cat} <unit>`? Arguably that would be the correct information but perhaps a bit misleading to the unaware?

**Comment by andir** [comment]
How shall we continue here? Since the reviews have so far been mostly around minor details I am hesitant to put in more work until we hear from @keszybz on the general approach.

**Comment by ReillyBrogan** [comment]
It would be great to see this one get merged. Solus Linux has a `/usr/lib` -> `/usr/lib64` symlink and has major issues without this patch.

File: src/basic/unit-file.c
Comment (keszybz) [reply src/basic/unit-file.c]: I think we need both. I'm rebasing the patch now and I added an comment that describes this.

File: src/basic/unit-file.c
Comment (yuwata) [inline src/basic/unit-file.c:442]: `resolved_dir` is leaked here.

Comment (keszybz) [reply src/basic/unit-file.c:442]: Arrggh!

File: src/basic/unit-file.c
Comment (bluca) [inline src/basic/unit-file.c]: /alias points is/alias is/

**Review by bluca — APPROVED** [review]

**Comment by keszybz** [comment]
CI passed. I'll update with the typo fixed. Is it fine to merge this? If yes, I don't think we need for the CI again.

**Comment by bluca** [comment]
It looks good to me, not sure if you were waiting for a review from @poettering too

**Review by DaanDeMeyer — APPROVED** [review]

**Pull request merged**
